### ISSUE TITLE
nn.MultiHeadAttention with different similarity measures

### ISSUE THREAD

**Issue by vainaijr** [pytorch/pytorch#31277]
## Pitch

Current nn.MultiHeadAttention uses matrix multiplication similarity, i.e., (Q@K.t()), but variants of this similarity are not available directly, for example, dot product, I.e., (Q*K.t()), additive similarity, (Wq*q.t() + Wk*k.t()), general dot product, (Q*W*K.t()).

These variants should also be in PyTorch.

**Comment by zou3519** [comment]
cc @zhangguanheng66 what do you think?

**Comment by zhangguanheng66** [comment]
Good point. We plan to break the long MultiheadAttention function into multiple building blocks. In that case, you could replace the scaled-product part with a custom one.
